core/src/dealer/mod.rs

```diff
@@ -0,0 +1,533 @@
+mod maps;
+mod protocol;
+
+use std::convert::TryFrom;
+use std::iter;
+use std::pin::Pin;
+use std::sync::atomic::AtomicBool;
+use std::sync::{atomic, Arc, Mutex};
+use std::task::Poll;
+
+use futures_core::{FusedFuture, Future, Stream};
+use futures_util::future::Fuse;
+use futures_util::{FutureExt, SinkExt, StreamExt};
+use tokio::select;
+use tokio::sync::mpsc::UnboundedReceiver;
+use tokio::sync::{mpsc, Semaphore};
+use tokio::task::JoinHandle;
+use tokio_tungstenite::tungstenite;
+use tungstenite::error::UrlError;
+use url::Url;
+
+use self::maps::*;
+use self::protocol::*;
+pub use self::protocol::{Message, Request};
+use crate::socket;
+use crate::util::keep_flushing;
+
+type WsMessage = tungstenite::Message;
+type WsError = tungstenite::Error;
+type WsResult<T> = Result<T, tungstenite::Error>;
+
+// TODO: better splitting
+fn split_uri(s: &str) -> impl Iterator<Item = &'_ str> {
+    if let Some(rest) = s.strip_prefix("hm://") {
+        iter::once("hm").chain(rest.split('/'))
+    } else if let Some(rest) = s.strip_suffix("spotify:") {
+        iter::once("spotify").chain(rest.split(':'))
+    } else {
+        panic!()
+    }
+    .filter(|x| !x.is_empty())
+}
```

Comment:
That's of course not acceptable as a long term solution.

I figured out that there are basically two kinds of uris, `spotify:*` and `hm://*`, and tried to put them in a "sub-uri/parent-uri` hierarchy by comparing the result of this function.

Is there a way to create this hierarchy without hardcoding the 2 kinds of uris?

Reply:
So seems to me like there are two kinds of formats -- for now at least. One is coded like URI's like `{scheme}://{slash-separated-path}`  and the other is `{scheme}:{colon-separated-path}`. Does that help in making it generic?

To do the one or the other you could maybe do something like `s.hyper::parse::<Uri>().unwrap_or_else()`

Reply:
According to [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986) it's not specified how a path can be interpreted hierarchically, it depends on the scheme. So maybe this solution is good enough, but it shouldn't panic if it fails